source/mir/las/sum.d

```diff
@@ -1815,6 +1815,98 @@ nothrow unittest
+            assert(summator.isNaN() == r.isNaN());
+            assert(summator.isFinite() == r.isFinite() || r == -double.max && s == -double.infinity || r == double.max && s == double.infinity);
+            assert(summator.isInfinity() == r.isInfinity() || r == -double.max && s == -double.infinity || r == double.max && s == double.infinity);
+            assert(nextDown(s) <= r && r <= nextUp(s) || s.isNaN && r.isNaN);
+        }
+        else
+        {
+            assert(summator.isNaN() == r.isNaN());
+            assert(summator.isFinite() == r.isFinite());
+            assert(summator.isInfinity() == r.isInfinity());
+            assert(s == r || s.isNaN && r.isNaN);
+        }
+    }
+}
+
+nothrow unittest
+{
+    import std.range: iota, retro;
+    import std.algorithm: map;
+    import std.array: array;
+
+    Summator!(double, Summation.kahan) summator = 0.0;
+
+    enum double M = (cast(double)2) ^^ (double.max_exp - 1);
+    Tuple!(double[], double)[] tests = [
+        tuple(new double[0], 0.0),
+        tuple([0.0], 0.0),
+        tuple([double.max, double.max*2.^^-54], double.max),
+        tuple([double.max, double.max*2.^^-53], double.infinity),
+        tuple(iota(1, 1001).map!(a => 1.0/a).array , 7.4854708605503451),
+        tuple(iota(1, 1001).map!(a => (-1.0)^^a/a).array, -0.69264743055982025), //0.693147180559945309417232121458176568075500134360255254120680...
+        tuple(iota(1, 1001).map!(a => 1.0/a).retro.array , 7.4854708605503451),
+        tuple(iota(1, 1001).map!(a => (-1.0)^^a/a).retro.array, -0.69264743055982025),
+        tuple([double.infinity, -double.infinity, double.nan], double.nan),
+        tuple([double.nan, double.infinity, -double.infinity], double.nan),
+        tuple([double.infinity, -double.infinity], double.nan),
+        tuple([M-2.0^^970, 0.0, M], double.infinity),
+        tuple([M-2.0^^970, 1.0, M], double.infinity),
+        tuple([M, M], double.infinity),
+        tuple([-M, 2.^^971, -M], -double.max),
+    ];
+    foreach (i, test; tests)
+    {
+        foreach (t; test[0]) summator.put(t);
+        auto r = test[1];
+        auto s = summator.sum;
+        version(X86)
+        {
+            assert(summator.isNaN() == r.isNaN());
+            assert(summator.isFinite() == r.isFinite() || r == -double.max && s == -double.infinity || r == double.max && s == double.infinity);
+            assert(summator.isInfinity() == r.isInfinity() || r == -double.max && s == -double.infinity || r == double.max && s == double.infinity);
+            assert(nextDown(s) <= r && r <= nextUp(s) || s.isNaN && r.isNaN);
+        }
+        else
+        {
+            assert(summator.isNaN() == r.isNaN());
+            assert(summator.isFinite() == r.isFinite());
+            assert(summator.isInfinity() == r.isInfinity());
+            assert(s == r || s.isNaN && r.isNaN);
+        }
+        summator = 0.0;
+    }
+}
+
+nothrow unittest
+{
+    // unittest with different result to precise summation
```

Comment:
Unite this test for kahan with previous one and generalize it with all except `precise`